Feature bjs2 6411 projects

Th0rn-dev opened the pull request:
Не стал покрывать тестами все методы ProjectService (пока еще буксую на сложных методах сервиса проекта)

Th0rn-dev committed:
initiated ProjectController

Th0rn-dev committed:
added ProjectService and refactored ProjectController and ProjectRepository.

Th0rn-dev committed:
refactored ProjectService and ProjectController

Th0rn-dev committed:
Merge branch 'werewolf-master-bc4' into feature-BJS2-6411_projects

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/ProjectController.java
#	src/main/java/faang/school/projectservice/service/ProjectService.java

Th0rn-dev committed:
added test for ProjectService and refactored ProjectService, ProjectController and mappers

Th0rn-dev committed:
refactored ProjectValidator
Combined validation for the project into one method and refactored methods create() of ProjectService
Refactored ProjectValidatorTest.

Th0rn-dev committed:
refactored update method

EKoregin commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -18,9 +30,53 @@ public class ProjectService {
+    public ProjectDto create(ProjectDtoRequest projectDtoRequest) {
+        Project project = projectMapper.requestDtoToProject(projectDtoRequest);
+        projectValidator.validateProjectParams(project);
+        log.warn("The owner ID does not match the authorized user ID.");
```
Лишнее сообщение.

Th0rn-dev replied:
Убрал

EKoregin commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -18,9 +30,53 @@ public class ProjectService {
+    public ProjectDto update(Long projectId, ProjectDto projectDto) {
+        projectValidator.isExists(projectId);
+        Project project = projectRepository.getProjectById(projectId);
+        ;
```
Потерялось двоеточие :)

Th0rn-dev replied:
Убрал

EKoregin commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -18,9 +30,53 @@ public class ProjectService {
+    @Transactional
+    public ProjectDto create(ProjectDtoRequest projectDtoRequest) {
+        Project project = projectMapper.requestDtoToProject(projectDtoRequest);
+        projectValidator.validateProjectParams(project);
```
Ты здесь отправляешь на валидацию поля сущности Project, которую создал через маппер из Dto. Лучше проверить поля в самой Dto, без использования валидатора и предварительного маппинга. В дтошке можно использовать аннотации для валидации, и в контроллере добавить для нее @Valid.

Th0rn-dev replied:
На самом деле requestDtoToProject(projectDtoRequest) много чего происходит: устанавливаются поля по умолчанию, определяется пользователь.
Мне так представляется, что сущность наглядней валидировать, чем добавлять дополнительные поля в DTO

Th0rn-dev replied:
Про аннотацию @Valid почитаю, спасибо.

EKoregin commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,34 @@
+
+    public void validateProjectParams(Project project) {
+        boolean isUniq = projectRepository.existsByOwnerUserIdAndName(project.getOwnerId(), project.getName());
+        if (Objects.isNull(project.getName()) || project.getName().isEmpty()) {
+            throw new DataValidationException("Project name must not be null or empty.");
+        } else if (Objects.isNull(project.getDescription()) || project.getDescription().isEmpty()) {
```
Можно проверять эти поля через валидацию в Dto

Th0rn-dev replied:
Отрефакторил валидатор и перенес проверку name и description в DTO

EKoregin commented on src/test/java/faang/school/projectservice/service/ProjectServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Test
+    public void testCreateProjectIsOwnerPresent() {
+        when(projectRepository.save(PROJECT)).thenReturn(PROJECT);
+        when(projectMapper.projectToDto(PROJECT)).thenReturn(PROJECT_DTO);
+        var project = projectRepository.save(PROJECT);
+        var dto = projectMapper.projectToDto(PROJECT);
+        var ProjectDto = projectService.create(PROJECT_CREATE_DTO);
+
+        verify(projectRepository, times(1)).save(PROJECT);
+        verify(projectMapper, times(1)).projectToDto(PROJECT);
+
+        assertEquals(PROJECT_DTO.getName(), project.getName());
+        assertEquals(PROJECT_DTO.getDescription(), project.getDescription());
+        assertEquals(ProjectStatus.CREATED, project.getStatus());
```
В строке 38 ты эмулируешь работу метода save и возвращаешь PROJECT.
В строке 40 ты получаешь тот же объект который возвращает save.
Далее в строках 47, 48, 49 ты проверяешь поля. Смысл проверки теряется, потому что ты проверяешь заведомо одинаковый объект.

Тебе в тесте нужно проверить конечный результат, а это ProjectDto, который ты получаешь в результате работы метода create() в сервисе.  Проверять замоканные объекты не нужно.

Также советую не использовать var, если не понятно какого типа возвращается переменная (при ревью кода это не очевидно).

Th0rn-dev replied:
Полностью переделал код теста testCreateProjectIsOwnerPresent()

Th0rn-dev committed:
refactored code by recommendations of cross review

EKoregin commented on src/test/java/faang/school/projectservice/service/ProjectServiceTest.java:
```diff
@@ -0,0 +1,94 @@
+        verify(projectMapper, times(1)).projectToDto(any(Project.class));
+        verify(projectMapper, times(1)).requestDtoToProject(any(ProjectDtoRequest.class));
+
+        assertNotNull(resultDto);
+        assertEquals(resultDto.getName(), PROJECT_DTO_REQUEST.getName());
+        assertEquals(resultDto.getDescription(), PROJECT_DTO_REQUEST.getDescription());
+        assertEquals(resultDto.getOwnerId(), resultDto.getOwnerId());
```
Эти проверки можно заменить одной, сравнив два объекта.

EKoregin replied:
assertEquals(resultDto.getOwnerId(), resultDto.getOwnerId()); - идет сравнение одного и того же :)

Th0rn-dev replied:
Отрефакторил - поменял местами ожидаемое и актуально

EKoregin commented on src/main/java/faang/school/projectservice/dto/project/ProjectDtoRequest.java:
```diff
@@ -0,0 +1,24 @@
+@Builder
+public class ProjectDtoRequest {
+    @Min(value = 1, message = "ID cannot be less than or equal to 0")
+    @NotBlank
```
Можно добавить понятное сообщение, которое увидит клиент

Th0rn-dev replied:
Поменял сообщение

EKoregin commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -1,31 +1,76 @@
 @RequestMapping("/projects")
 @RequiredArgsConstructor
-@Validated
+@Valid
```
Аннотация @Valid ставится перед ДТОшкой, которая передается в контроллер в кач-ве параметра после @RequestBody.
@Validated ставится над классом, чтобы происходили проверки, которые добавлены для отдельных параметров в контроллере.

Th0rn-dev replied:
Отрефакторил

Th0rn-dev replied:
Переставил аннотацию и увидел, что твою удалил. Поправил

Th0rn-dev committed:
refactored code by recommendations of cross review 2

Th0rn-dev committed:
Merge branch 'werewolf-master-bc4' into feature-BJS2-6411_projects

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/ProjectController.java

Th0rn-dev committed:
merged with werewolf-master-bc4

Th0rn-dev committed:
Merge branch 'werewolf-master-bc4' into feature-BJS2-6411_projects

InnMish commented on src/main/java/faang/school/projectservice/filter/project/ProjectNameFilter.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public class ProjectNameFilter implements ProjectFilter{
```
А ты тестировал это? У тебя сейчас фильтры не будут работать) 
Прогони в постмане запрос и проверь фильтрацию

Нужно чинить

Th0rn-dev replied:
Добавил аннотацию @Component. 
Немного отрефакторил и покидал запросы с фильтрацией из Postman

InnMish commented on src/main/java/faang/school/projectservice/dto/project/ProjectDtoRequest.java:
```diff
@@ -0,0 +1,24 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class ProjectDtoRequest {
```
А зачем тебе вторая дто? Почему нельзя тут ProjectDto использовать для создания? Просто статус проставлять нужный в коде. Не вижу нужды в том классе

Th0rn-dev replied:
Удалил ProjectDtoRequest и утильный класс

InnMish commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Mapping(target = "status", constant = "CREATED")
+    @Mapping(target = "visibility", defaultValue = "PRIVATE")
+    @Mapping(target = "ownerId", qualifiedByName = {"UserMapperUtil", "setOwner"}, source = "ownerId")
```
UserMapperUtil это что? почему ты таким образом owner сетишь? 
Ты это поле из дто можешь достать, зачем ютил?

Th0rn-dev replied:
Тут условие по задаче. Если ownerId не установлен, сетим id залогиненного юзера

Th0rn-dev replied:
Засетил поля в ProjectService

InnMish commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,35 @@
+    @Mapping(target = "ownerId", qualifiedByName = {"UserMapperUtil", "setOwner"}, source = "ownerId")
+    Project requestDtoToProject(ProjectDtoRequest projectDtoRequest);
+
+    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
```
этк аннотацию ты можешь не на метод навесить, а на класс весь 
Чтобы все null поля игнорились, а не только для этого метода

Th0rn-dev replied:
Навесил на класс

InnMish commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,35 @@
+    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+    Project dtoToProject(ProjectDto projectDto, @MappingTarget Project entity);
+
+    ProjectDto projectToDto(Project project);
```
метод называй просто toEntity и toDto 
Мы по принимаемому и возвращаемому типу уже и так видим, с чем работаем

Th0rn-dev replied:
Переименовал

InnMish commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -1,35 +1,98 @@
 
     @Transactional
-    public void uploadFile(MultipartFile file, Long projectId, String folder) {
+    public ProjectDto update(@Min(1) Long projectId, ProjectDto projectDto) {
```
@Min(1) валидацию на уровень контроллера можно вынести

Th0rn-dev replied:
вынес валидацию в DTO

InnMish commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -1,35 +1,98 @@
     @Transactional
-    public void uploadFile(MultipartFile file, Long projectId, String folder) {
+    public ProjectDto update(@Min(1) Long projectId, ProjectDto projectDto) {
+        projectValidator.isExists(projectId);
```
метод назови просто exists

InnMish commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -1,35 +1,98 @@
+    }
+
+    @Transactional(readOnly = true)
+    public ProjectDto findById(@Min(1) Long id) {
```
валидацию на уровень контроллера

InnMish commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -1,35 +1,98 @@
+        return projectMapper.projectsToDtos(projectRepository.findAll());
+    }
+
+    public void delete(@Min(1) Long projectId) {
```
валидацию на уровень контроллера

InnMish commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -1,35 +1,98 @@
         projectRepository.save(project);
     }
+
+    private boolean isUserExistInTeams(Team team) {
```
назови метод isTeamMember

Th0rn-dev replied:
Переименовал

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public void validateProjectIsUniqByIdAndName(Project project) {
+        boolean isUniq = projectRepository.existsByOwnerUserIdAndName(project.getOwnerId(), project.getName());
```
поле назови unique

Th0rn-dev replied:
Назвал

InnMish commented on src/main/resources/db/changelog/changeset/project_V014_update_data.sql:
```diff
@@ -0,0 +1,17 @@
+INSERT INTO project (name, description, owner_id, status, visibility)
+VALUES ('Market', 'Market description', 1, 'CREATED', 'PUBLIC');
+
+UPDATE project
```
зачем тебе это в миграции? Миграция изменяет данные в бд, которые потом остаются такими 
Если ты что-то тетстируешь, тебе нужно отправлять динамические запросы, то есть поднять приложение и затем в реальзном времени к запущенному приложению отправлять update, если тебе нужно это проверить. 

В миграции мы только создаем таблицы. Или добавлям/удаляем в таблицу какие-то колонки. Но не работаем конкретно с записями в таблице

Th0rn-dev replied:
Удалил миграцию.
У нас БД пустая, хотел таким образом сохранить несколько проектов.
Сделаю себе скрипт, чтобы пулить данные, чтобы с ними работать.

InnMish commented on src/main/java/faang/school/projectservice/mapper/ProjectMapperUtil.java:
```diff
@@ -0,0 +1,23 @@
+@Named("UserMapperUtil")
+@Component
+@RequiredArgsConstructor
+public class ProjectMapperUtil {
```
почему ты это здесь делаешь? 
Если тебе нужно из другого сервиса достать айди, делай это не в маппере, а в сервисе. Маппер отвечает ТОЛЬКО за преобразование одной сущности в другую. Он не занимается межсервисным взаимодейтсвием

Th0rn-dev replied:
Удалил утильный класс, логику обрабатываю в сервисе

Th0rn-dev committed:
refactored ProjectService and ProjectController

Th0rn-dev committed:
fixed problem of ambiguous choice
Solution to the problem of ambiguous choice of mappers when adding an interface

Th0rn-dev committed:
refactored ProjectValidator and ProjectValidatorTest

Th0rn-dev committed:
refactored ProjectController and ProjectService + test

Th0rn-dev committed:
added ProjectNameFilterTest

Th0rn-dev committed:
Merge branch 'werewolf-master-bc4' into feature-BJS2-6411_projects

Th0rn-dev committed:
fixed the problem with the filter

Th0rn-dev committed:
added updateEntity to ProjectMapper and refactored ProjectService, ProjectServiceTest

Th0rn-dev committed:
added INFO.md
